x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/regression/RSquared.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.core.ml.dataframe.evaluation.regression;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.search.aggregations.AggregationBuilder;
+import org.elasticsearch.search.aggregations.AggregationBuilders;
+import org.elasticsearch.search.aggregations.Aggregations;
+import org.elasticsearch.search.aggregations.metrics.ExtendedStats;
+import org.elasticsearch.search.aggregations.metrics.ExtendedStatsAggregationBuilder;
+import org.elasticsearch.search.aggregations.metrics.NumericMetricsAggregation;
+import org.elasticsearch.xpack.core.ml.dataframe.evaluation.EvaluationMetricResult;
+
+import java.io.IOException;
+import java.text.MessageFormat;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Calculates R-Squared between two known numerical fields.
+ *
+ * equation: R-Squared = 1 - SSres/SStot
+ * such that,
+ * SSres = Σ(y - y´)^2, The residual sum of squares
+ * SStot =  Σ(y - y_mean)^2, The total sum of squares
+ */
+public class RSquared implements RegressionMetric {
+
+    public static final ParseField NAME = new ParseField("r_squared");
+
+    private static final String PAINLESS_TEMPLATE = "def diff = doc[''{0}''].value - doc[''{1}''].value;return diff * diff;";
+    private static final String SS_RES = "residual_sum_of_squares";
+
+    private static String buildScript(Object...args) {
```

Comment:
nit: why make this a varargs parameter when we know the template takes 2 replacements

Reply:
This allows the call down to the `MessageFormat#format` to be done without warnings or errors and no explicit casting. using a `String var1, String var2` requires casting to get around warnings, and using `Object var1, Object var2` requires manual construction of an array so that it matches the appropriate function definition.